saml/src/main/java/com/linecorp/armeria/saml/HttpRedirectBindingUtil.java

```diff
@@ -0,0 +1,292 @@
+
+    /**
+     * HTTP parameter names which are used for HTTP-Redirect binding protocol.
+     */
+    static final String PARAM_SAML_REQUEST = "SAMLRequest";
+    static final String PARAM_SAML_RESPONSE = "SAMLResponse";
+    static final String PARAM_SIGNATURE = "Signature";
+    static final String PARAM_SIGNATURE_ALGORITHM = "SigAlg";
+    static final String PARAM_RELAY_STATE = "RelayState";
+
+    private static final String DEFAULT_CC =
+            String.join(",", HttpHeaderValues.NO_CACHE, HttpHeaderValues.NO_STORE);
+    private static final String DEFAULT_PRAGMA = HttpHeaderValues.NO_CACHE.toString();
+
+    /**
+     * Returns an {@link HttpHeaders} with the specified {@code location}, a default {@code cache-control}
+     * and a default {@code pragma} headers.
+     */
+    static HttpHeaders headersWithLocation(String location) {
+        return HttpHeaders.of(HttpStatus.FOUND)
+                          .add(HttpHeaderNames.LOCATION, location)
+                          .add(HttpHeaderNames.CACHE_CONTROL, DEFAULT_CC)
+                          .add(HttpHeaderNames.PRAGMA, DEFAULT_PRAGMA);
+    }
+
+    /**
+     * Returns an {@link HttpResponse} with the specified {@code location}, a default {@code cache-control}
+     * and a default {@code pragma} headers.
+     */
+    static HttpResponse responseWithLocation(String location) {
+        return HttpResponse.of(headersWithLocation(location));
+    }
+
+    /**
+     * Returns a redirected URL that the specified SAML message is deflated and base64-encoded.
+     * The URL must contain a signature of the generated query string.
+     */
+    static String toRedirectionUrl(SAMLObject msg,
+                                   String endpointUrl,
+                                   String messageParamName,
+                                   Credential signingCredential,
+                                   String signatureAlgorithm,
+                                   @Nullable String relayState) {
+        requireNonNull(msg, "msg");
+        requireNonNull(endpointUrl, "endpointUrl");
+        requireNonNull(messageParamName, "messageParamName");
+        requireNonNull(signingCredential, "signingCredential");
+        requireNonNull(signatureAlgorithm, "signatureAlgorithm");
+
+        try {
+            final QueryStringEncoder encoder = new QueryStringEncoder(endpointUrl);
+
+            final String samlRequest = deflateAndBase64Encode(msg);
+            encoder.addParam(messageParamName, samlRequest);
+
+            if (relayState != null) {
+                if (relayState.length() > 80) {
+                    logger.warn("too long relayState string: " + relayState.length());
+                }
+                encoder.addParam(PARAM_RELAY_STATE, relayState);
+            }
+
+            encoder.addParam(PARAM_SIGNATURE_ALGORITHM, signatureAlgorithm);
+
+            // Use URL-encoded query string as input.
+            final String input = encoder.toUri().getRawQuery();
```

Comment:
I think you can use `encoder.toString()` (maybe need substring if there's a `?` but still much more efficient than going through the URI parser)